### PR TITLE
Simplify caching using ES6 Map()

### PR DESCRIPTION
**Pull request by semenko**
With EMCA2015, we can use a simple Map() instead of the more complex LRU engine.

This is no longer an LRU (now a FIFO), but seems to be more performant -- and it's certainly cleaner.

(+ minor cleanups.)

### CODE REVIEW

**Comment by semenko** [comment]
@jsha for review, unless anyone else wants to look at the chrome code :)

**Comment by jsha** [comment]
I definitely like the simplification. It's interesting that `Map.prototype.keys()` returns keys in insertion order rather than random or alphabetically. I didn't know that! It's probably worth commenting about it on the line where you remove.

You said it seems to be more performant. Is that based on performance testing, or general feel? If the former, can you share your methodology?

File: chromium/rules.js [hunk 1]
```diff
@@ -287,6 +282,10 @@ RuleSets.prototype = {
 
     // Insert results into the ruleset cache
     this.ruleCache.set(host, results);
+
+    // Cap the size of the cache. (Limit chosen somewhat arbitrarily)
+    if (this.ruleCache.size > 1000) { this.ruleCache.delete(this.ruleCache.keys().next().value) }
```
Comment (jsha) [inline chromium/rules.js]: Newline after the brace, please.

File: chromium/rules.js [hunk 2]
```diff
@@ -357,6 +356,9 @@ RuleSets.prototype = {
     var nonce_path = "/" + Math.random().toString();
     var test_uri = "http://" + domain + nonce_path + nonce_path;
 
+    // Cap the size of the cookie cache (limit chosen somewhat arbitrarily)
+    if (this.cookieHostCache.size > 250) { this.cookieHostCache.delete(this.cookieHostCache.keys().next().value) }
```
Comment (jsha) [inline chromium/rules.js]: Same here.

**Comment by semenko** [comment]
Thanks for the review! Making those changes and will rebase after you merge #3961 since they'll conflict on a line or two.

Alas, performant is based on a squishy "pseudo-testing" with Chrome's Javascript CPU profiler while opening sets of tabs (in a fresh profile) and observing the % cpu & timing info.

I'd love a better way to do this programmatically, or in Travis -- any thoughts?

**Comment by jsha** [comment]
Another question: I think this is the first hard dependency on ECMA2015 in the Chrome code, right?

I would like, at some point not too far off, to move the rules code into a shared directory, and migrate the Firefox extension to it, so we can get the benefit of these recent performance improvements (and generally cleaner code). We need to support Firefox ESR for the purposes of the Tor browser, and well as Firefox latest. If you could investigate whether `Map()` and similar ECMA2015 features are available in ESR, I'd greatly appreciate it. If they're not, I'd prefer to use other functionality where possible.

**Comment by semenko** [comment]
Yep -- this is the first hard ES6 dependency, and it's compatible with FF38 ESR, per: https://kangax.github.io/compat-table/es6/

I commented a bit in #2640 too -- I'm trying to clean the Chrome implementation and stick to APIs (unrelated to this PR) that are supported by FF's new WebExtension model.

**Comment by jsha** [comment]
LGTM. Needs merge.

**Comment by semenko** [comment]
Merging per above LGTM.

**Comment by RReverser** [comment]
This broke [rewriter](https://github.com/EFForg/https-everywhere/issues/2926#issuecomment-181845601) that is used for various tests.

**Comment by semenko** [comment]
Cool -- any thoughts on fixing it? (I'm not that familiar with how it works -- though it looks like you could just rip out the lrucache reference.)

**Comment by RReverser** [comment]
@semenko Looks like pretty much that's it (plus fix RuleSets constructor - with new approach it should receive smth like `Object.create(null)` by default?). This fix should be easy, but more generally, as mentioned in that comment, would be nice to have separate parts of HTTPS Everywhere decoupled so that they could be used and updated independently without future code breakages. Do you think that would be possible?

**Comment by semenko** [comment]
Yeah, it's definitely something good to do, though we'll need think about the best way to do it while preserving git history, etc.

**Comment by RReverser** [comment]
@semenko Git history can be rewritten AFAIK - e.g. you can see entire UNIX history converted to Git few days ago https://github.com/dspinellis/unix-history-repo/commits/master?page=3951

**Pull request merged**